bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java

```diff
@@ -0,0 +1,181 @@
+    private static final Color COLOR_PANEL = Color.BLACK;
+    private static final Color COLOR_SIDE = Color.GRAY;
+    private static final Color COLOR_TEXT = Color.BLACK;
+
+    public static byte[] render(PanelLayout panelLayout) throws IOException {
+        double rotationRadians = 0;
+        GlobalOrientation globalOrientation = panelLayout.getGlobalOrientation();
+        if (globalOrientation != null) {
+            rotationRadians = calculateRotation(globalOrientation);
+        }
+
+        Layout layout = panelLayout.getLayout();
+        if (layout == null) {
+            return new byte[] {};
+        }
+
+        List<PositionDatum> panels = layout.getPositionData();
+        if (panels == null) {
+            return new byte[] {};
+        }
+
+        Point2D size[] = findSize(panels, rotationRadians);
+        final Point2D min = size[0];
+        final Point2D max = size[1];
+        Point2D prev = null;
+        Point2D first = null;
+
+        int sideCounter = 0;
+        BufferedImage image = new BufferedImage(
+                (max.getX() - min.getX()) + 2 * NanoleafBindingConstants.LAYOUT_BORDER_WIDTH,
+                (max.getY() - min.getY()) + 2 * NanoleafBindingConstants.LAYOUT_BORDER_WIDTH,
+                BufferedImage.TYPE_INT_RGB);
+        Graphics2D g2 = image.createGraphics();
+
+        g2.setBackground(COLOR_BACKGROUND);
+        g2.clearRect(0, 0, image.getWidth(), image.getHeight());
+
+        for (PositionDatum panel : panels) {
+            final ShapeType shapeType = ShapeType.valueOf(panel.getShapeType());
+
+            Shape shape = ShapeFactory.CreateShape(shapeType, panel);
+            List<Point2D> outline = toPictureLayout(shape.generateOutline(), image.getHeight(), min, rotationRadians);
+            for (int i = 0; i < outline.size(); i++) {
+                g2.setColor(COLOR_SIDE);
+                Point2D pos = outline.get(i);
+                Point2D nextPos = outline.get((i + 1) % outline.size());
+                g2.drawLine(pos.getX(), pos.getY(), nextPos.getX(), nextPos.getY());
+            }
+
+            for (int i = 0; i < outline.size(); i++) {
+                Point2D pos = outline.get(i);
+                g2.setColor(COLOR_PANEL);
+                g2.fillOval(pos.getX() - NanoleafBindingConstants.LAYOUT_LIGHT_RADIUS / 2,
+                        pos.getY() - NanoleafBindingConstants.LAYOUT_LIGHT_RADIUS / 2,
+                        NanoleafBindingConstants.LAYOUT_LIGHT_RADIUS, NanoleafBindingConstants.LAYOUT_LIGHT_RADIUS);
+            }
+
+            Point2D current = toPictureLayout(new Point2D(panel.getPosX(), panel.getPosY()), image.getHeight(), min,
+                    rotationRadians);
+            if (sideCounter == 0) {
+                first = current;
+            }
+
+            g2.setColor(COLOR_SIDE);
+            final int expectedSides = shapeType.getNumSides();
+            if (shapeType.getDrawingAlgorithm() == DrawingAlgorithm.CORNER) {
```

Comment:
I only now discover that you use the .CORNER type as well (you'll see other comments below because I only review this part as the last piece of code after having gone through all the others). However, I don't understand what the corner type is for. Can you elaborate on that or even document it in the code?

Reply:
Good idea! Added a comment. They are very different from the traditional shapes.